apps/browser/src/vault/popup/services/vault-popup-scroll-position.service.ts

```diff
@@ -0,0 +1,77 @@
+import { CdkVirtualScrollableElement } from "@angular/cdk/scrolling";
+import { inject, Injectable } from "@angular/core";
+import { takeUntilDestroyed } from "@angular/core/rxjs-interop";
+import { NavigationEnd, Router } from "@angular/router";
+import { filter, skip, Subscription } from "rxjs";
+
+@Injectable({
+  providedIn: "root",
+})
+export class VaultPopupScrollPositionService {
+  private router = inject(Router);
+
+  /** Path of the vault screen */
+  private readonly vaultPath = "/tabs/vault";
+
+  /** Current scroll position relative to the top of the viewport. */
+  private scrollPosition: number | null = null;
+
+  /** Subscription associated with the virtual scroll element. */
+  private scrollSubscription: Subscription | null = null;
+
+  constructor() {
+    this.router.events
+      .pipe(
+        takeUntilDestroyed(),
+        filter((event): event is NavigationEnd => event instanceof NavigationEnd),
+      )
+      .subscribe((event) => {
+        this.resetListenerForNavigation(event);
+      });
+  }
+
+  /** Scrolls the user to the stored scroll position and starts tracking scroll of the page. */
+  start(virtualScrollElement: CdkVirtualScrollableElement) {
+    if (this.scrollPosition) {
+      // Use `setTimeout` to scroll after rendering is complete
+      setTimeout(async () => {
+        // `?? 0` is only to make typescript happy. It shouldn't happen with the above truthy check for `this.scrollPosition`.
+        virtualScrollElement.scrollTo({ top: this.scrollPosition ?? 0, behavior: "instant" });
```

Comment:
⛏️ (non-blocking): We can also use the [non-null assertion operator (`!`)](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator) and avoid the `?? 0` and explanatory comment.
```suggestion
        virtualScrollElement.scrollTo({ top: this.scrollPosition!, behavior: "instant" });
```

Reply:
[bf707cd](https://github.com/bitwarden/clients/pull/12325/commits/bf707cdcd58513eb89149415237f3b56b3fb2482)